design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,486 @@
+platform:
+  baremetal:
+    networkPolicy:
+      ipPools:
+        traffic-1:
+          cidr: 10.10.10.0/24
+          start: 10.10.10.1
+          end: 10.10.10.10
+          exclude: ["10.10.10.4", "10.10.10.7"]
+        traffic-2:
+          cidr: 10.10.11.0/24
+          start: 10.10.11.1
+          end: 10.10.10.11.10
+          exclude: ["10.10.11.4", "10.10.11.7"]
+      capture:
+        address-from-traffic-1: ippool.traffic-1.allocate
+        address-from-traffic-2: ippool.traffic-2.allocate
+      desiredState:
+        interfaces:
+        - name: eth1
+          type: ethernet
+          state: up
+          ipv4:
+            address:
+            - ip: "{{ capture.address-from-foo.ip }}"
+              prefix-length: "{{ capture.address-from-foo.prefix-length }}"
+            dhcp: false
+            enabled: true
+        - name: eth2
+          type: ethernet
+          state: up
+          ipv4:
+            address:
+            - ip: "{{ capture.address-from-bar.ip }}"
+              prefix-length: "{{ capture.address-from-bar.prefix-length }}"
+            dhcp: false
+            enabled: true
+    hosts:
+      - name: openshift-master-0
+        networkData: [static network configuration for master-0]
+      - name: openshift-worker-0
+        networkData: [static network configuration for worker-0]
+```
+
+The new field `platform.baremetal.networkPolicy` will contain the
+content of the day 2 `NodeNetworkConfigurationPolicy` and also the `ipPools`
+since there is not going to be multiple `networkPolicy` fields.
+
+### Implementation Details/Notes/Constraints
+
+#### Day 2
+
+Since kubernetes-nmstate has a distributed architecture each of the handler
+pods will calculate an allocatable ip from the `NodeIPPool` status and
+retry the process on conflict
+
+Handlers will check NNCEs at Reconcile and if the ip has being already assigned
+it will re-use that and update `NodeIPPool` status with it, this way the
+`NodeIPPool` can be reset but IPs are kept.
+
+To optimize on storage and lookup instead of using a list of allocated offset
+it's possible to use a bitmap, the OVN project use that approach to store the
+offsets on a range [1], a snapshot of the bitmap can be stored as base64 at the
+`NodeIPPool` status.
+
+[1] https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ipallocator/allocator.go
+
+#### Baremetal IPI
+
+The `image-customize-controller` [1] is a golang controller so it would be quite
+easy to import ippool library from kubernetes-nmstate and nmpolicy to render the
+`networkPolicy`.
+
+[1] https://github.com/openshift/image-customization-controller/blob/3ad18911d14dde1c23b660f99bb05e12a21c99e4/pkg/ignition/builder.go#L79
+
+### Risks and Mitigations
+
+If a bug prevents IPs deallocation the ippool can be exhausted and end up
+with network reconfiguration or scaling up failures. To fix those situations
+the field `status.allocated-offsets` at `NodeIPPool` can be set with an
```

Comment:
> What is the trigger to reconstruct the pool? Empty `status.allocated-offsets`list? 

Yes

> How do we know the pool actually needs a reconstruction and it's not just the the list is empty because no address is used? 

If no address is used the `ippools` field at NNCE will be empty and nothing will be done

> What component is responsible to reconstruct the pool?

Is distributed across the handlers, every handler try to allocate an address for it's interfaces, we do an optimistic locking with retry on failure like we do at NNCP status.

Reply:
and the NNCP will be marced as not available.